### PR TITLE
ZXW Diagram implementation

### PR DESCRIPTION
**Pull request by exAClior**
Implement `ZXWDiagram`.
Moving functions under `zx_diagram.jl` to `utils.jl` for reusing them for `ZXWDiagram`.

### CODE REVIEW

**Comment by ChenZhao44** [comment]
Thank you for the pull request; it looks great overall. I have a few minor suggestions. It seems that you have extended the `SpiderType` and used it in the `ZXWDiagram`. This extension also makes the original `ZXDiagram` capable of handling ZXW, resulting in no difference between the new `ZXWDiagram`. I suggest defining a new spider type for `ZXWDiagram` only to avoid affecting the previous data structures and algorithms.

**Comment by exAClior** [comment]
> Thank you for the pull request; it looks great overall. I have a few minor suggestions. It seems that you have extended the `SpiderType` and used it in the `ZXWDiagram`. This extension also makes the original `ZXDiagram` capable of handling ZXW, resulting in no difference between the new `ZXWDiagram`. I suggest defining a new spider type for `ZXWDiagram` only to avoid affecting the previous data structures and algorithms.

This makes much more sense. I will change it accordingly. Thank you!

**Comment by ChenZhao44** [comment]
Since we want to separate the original implementation, we can also use algebraic data types to represent the node type and its parameters. An algebraic data type is a type that combines several data types in a type-stable way. An implementation of ADTs can be found in [Expronicon](https://expronicon.rogerluo.dev/intro/adts/intro). By using ADT pattern matching, as implemented in [MLStyle](https://thautwarm.github.io/MLStyle.jl/latest/syntax/pattern.html), we can replace the `if else` statements during matching calculus rules in an elegant way. We can discuss this in more detail during our next meeting.

**Comment by exAClior** [comment]
The latest commits on ADT of `Parameters` used to represent parameters for spider didn't pass the CI because `YaoHIR` and `CompilerPluginTools` are restricting `Expronicon` to be `0.6.0`. I have already made a [PR](https://github.com/QuantumBFS/YaoHIR.jl/pull/7#issue-1771436278) in `YaoHIR` to partially solve the problem. I am temporarily removing `CompilerPluginTools` from `Project.toml` since Roger said he wants to rewrite `CompilerPluginTools` instead of fixing it.

**Comment by exAClior** [comment]
Hello, @ChenZhao44  I have finished correcting for `ZXWDiagram`, and `Parameter` and copied from your branch the implementation of converting a `ZXWDiagram` to `eincode` and then to `Matrix`. I am waiting for @Roger-luo 's comment on my [edits](https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl/pull/10#issue-1774145100) on `CompilerPluginTools`. For the sake of passing CI, I have removed all tests related to `CompilerPluginTools`.

File: src/parameter.jl [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    PiUnit(_...) => (p.pu isa Number) && (-1)^p.pu > 0
+    Factor(_...) => p.f ≈ 1.0 # making meaning the same, upto machine precision
+end
+
+function add_param(p1, p2)
+    @match (p1, p2) begin
+        (PiUnit(pu1, _), PiUnit(pu2, _)) && if pu1 isa Number && pu2 isa Number
+        end => Parameter(Val(:PiUnit), pu1 + pu2)
+        (PiUnit(pu1, _), PiUnit(pu2, _)) && if !(pu1 isa Number) || !(pu2 isa Number)
+        end => Parameter(Val(:PiUnit), Expr(:call, :+, pu1, pu2))
```
Comment (ChenZhao44) [inline src/parameter.jl:113]: Here we should include the number type of the expr. See https://github.com/QuantumBFS/ZXCalculus.jl/blob/bbc19ab5ec0f1eb0b09ffa7a6e2976f60b625fbc/src/phase.jl#L30C35-L30C35

File: src/parameter.jl [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+Base.:(+)(p1::Parameter, p2::Number) = add_param(p1, p2)
+Base.:(+)(p1::Number, p2::Parameter) = add_param(p2, p1)
+
+function subt_param(p1, p2)
+    @match (p1, p2) begin
+        (PiUnit(pu1, pu_t1), PiUnit(pu2, pu_t2)) && if pu1 isa Number && pu2 isa Number
+        end => Parameter(Val(:PiUnit), pu1 - pu2)
+        (PiUnit(pu1, pu_t1), PiUnit(pu2, pu_t2)) &&
+            if !(pu1 isa Number) || !(pu2 isa Number)
+            end => Parameter(Val(:PiUnit), Expr(:call, :-, pu1, pu2))
```
Comment (ChenZhao44) [inline src/parameter.jl:137]: Same as above

**Review by ChenZhao44 — APPROVED** [review]
This PR looks great. I only have two minor suggestions. Once you fix this, you can merge it.

File: src/parameter.jl (hunk 1 above)
Comment (exAClior) [reply src/parameter.jl:113]: Thank you very much! I corrected the `Expr` part. For the pattern matched above that, I declared the constructor of `Parameter` as a parametric method. The type of `pu` should be automatically matched like in the previous implementation [link](https://github.com/QuantumBFS/ZXCalculus.jl/blob/bbc19ab5ec0f1eb0b09ffa7a6e2976f60b625fbc/src/phase.jl#L27C31-L27C31) The test passed alright, I will merge the changes for now.

File: src/parameter.jl (hunk 2 above)
Comment (exAClior) [reply src/parameter.jl:137]: Changed here too.

**Pull request merged**